src/charm.py

```diff
@@ -633,7 +669,66 @@ def _get_backup_key(self) -> str:
         timestamp = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
         return f"{BACKUP_KEY_PREFIX}-{self.model.name}-{timestamp}"
 
-    def _get_ca_cert_location_in_charm(self) -> str:
+    def _on_new_access_certificate_joined(self, event: RelationJoinedEvent):
+        if not self._ingress_address:
+            raise RuntimeError("Tried to process relation join before peer relation.")
```

Comment:
I'm not sure about raising here. This looks like a realistic code path. We have a race condition and in cases where the certificate joined event is triggered before the pair relation is triggered, the charm will raise which is highly undesirable. We likely should handle whatever's here in the `_configure` function instead with validations like
```python
if self._access_certificate_relation_available() and not self._certificates_requested():
    self.request_certificates()
```
This will make it so that we for every event observed, we make the validations and requests necessary.

Reply:
Yes, if it makes sense of course. This will make it so that for (almost) any event, we look at the state of the world, we look at the charm, we compare the two and make the appropriate changes. It also centralises status handling over there.

Reply:
The reason for raising a Runtime Error here is that this code path is not possible to reach under normal circumstances. This statement should never run. This is also true for the other function where we raise this error.

Is there a way to replicate having the tls-access-certificate relation fire before the peer relation does? 

As an aside;

I think using a central configuration function makes sense in a general way, but I feel like the benefits of using the contexts of the events that Juju and the relations fire and the benefits that come with having composable pieces of lifecycle management outweigh the convenience of having one very large function determine all of the charm state and circumvent the validation issues that arise with mypy not being able to identify the enforced order of juju events as per the graph [here](https://juju.is/docs/sdk/charm-lifecycle). Especially when it comes to this relation, where the things that are supposed to happen have almost no dependencies on the other state variables of the charm except the ingress address.

I think we're already running into issues with this approach, like the PKI stuff not being able to be worked on, since conflicts are now very difficult to avoid, and Juju's "commitment issues" for adding secrets and other changes back into the charm ecosystem before the _configure() function exits without error, and having a single point of failure when problems with the real world happen, like pod crashes and such.

I also think there could be a middle ground, where logical parts of the lifecycle could have their own configuration functions, like `configure_pebble()` and `configure_network()` and `configure_certificates()` and `configure_kv()` etc. and actions can stay outside of this paradigm entirely.

What do you think?

Reply:
Talked and agreed about where to go here on a call